pkg/stash/with_etcd.go

```diff
@@ -0,0 +1,95 @@
+package stash
+
+import (
+	"context"
+	"time"
+
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/clientv3/concurrency"
+	"golang.org/x/sync/errgroup"
+)
+
+// WithEtcd returns a distributed singleflight
+// using an etcd cluster. If it cannot connect,
+// then it will panic
+func WithEtcd(endpoints []string) (Wrapper, error) {
+	const op errors.Op = "stash.WithEtcd"
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	c, err := clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: time.Second * 5,
+	})
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	var eg errgroup.Group
+	for _, ep := range endpoints {
+		eg.Go(func() error {
+			_, err := c.Status(ctx, ep)
+			return err
+		})
+	}
+	err = eg.Wait()
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	return func(s Stasher) Stasher {
+		return &etcd{c, s}
+	}, nil
+}
+
+type etcd struct {
+	c *clientv3.Client
+	s Stasher
+}
+
+func (s *etcd) Stash(ctx context.Context, mod, ver string) error {
+	const op errors.Op = "etcd.Stash"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+
+	mv := config.FmtModVer(mod, ver)
+	sesh, err := concurrency.NewSession(s.c)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	mu := concurrency.NewMutex(sesh, mv)
+	err = mu.Lock(ctx)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	defer mu.Unlock(ctx)
+	resp, err := s.c.Get(ctx, mv)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	if resp.Count == 0 {
+		lease, err := s.c.Grant(ctx, 15*60)
+		if err != nil {
+			return errors.E(op, err)
+		}
+		err = s.s.Stash(ctx, mod, ver)
```

Comment:
What if this fails? Does that mean that every subsequent request for this particular mod@ver will fail until the lease expires and the key is cleared?

Reply:
yes and they will fail fast :)

Reply:
@marpio @michalpristas very valid point. I figured that's okay...but maybe not? If not, then I can do the Distributed WaitGroup I mentioned above and this issue would be gone, but it does bring complexity higher. We can also potentially have a `WithRetries` stasher that retries fetching from upstream only when unexpected errors happen (not when "not found" errors happen)

Reply:
I would really :heart: to see the the distributed WaitGroup implemented but imho the `WithRetries` stasher would suffice. Could you outline any others advantages of the WaitGroup, beside this issue and being able to clear the keys more efficiently?

Reply:
+1 on ❤️ distributed WaitGroup, but this looks like an awesome start to me. Are peeps ok with doing that as a follow up?

Reply:
Hey peeps, so I've rethought how to do this and potentially found a much simpler way...it's sorta cheating but I think in a good way. I've updated the code so feel free to do another round of review. But in a nutshell here's what happens: 

1. Request comes in 
2. Issue a lock based on mod/ver
3. The lock then checks if the mod/ver exists through `storage.Checker` 
4. If it exists, then "someone before me already did this", we're good. 
5. If it doesn't exist, call the underlying stasher, and be the first to "stash" this. 

The only catch here, is if a previous stasher error'd out, we're only gonna return that error back to the user and not spread it around...This wouldn't qualify as singleflight anymore, but I think it's actually more a feature than a bug because we get free retries. 

In other words: 

1. User A tries, and gets an error. 
2. User B tries 1 second after, but will get a retry instead of User A's error.